{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Genome and Injecting SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Generate Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# python imports\n",
    "from random import choice\n",
    "import random\n",
    "\n",
    "# Biopython imports\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import DNAAlphabet\n",
    "from Bio import SeqIO\n",
    "import glob\n",
    "\n",
    "def createFrag(size):\n",
    "    DNA = \"\"\n",
    "    for count in range(size * 1000):\n",
    "        DNA += choice(\"CGTA\")\n",
    "    return DNA\n",
    "\n",
    "def get_haps(heterozygous_rate, size_kB):\n",
    "    gene = createFrag(size_kB)\n",
    "    alpha = set(['A', 'C', 'G', 'T'])\n",
    "    records = []\n",
    "\n",
    "    s_len = len(gene)\n",
    "    s = gene\n",
    "    delta = round(s_len * heterozygous_rate)\n",
    "    edit_pos = random.sample(range(0, s_len), delta)\n",
    "    sl = list(gene)\n",
    "\n",
    "    \n",
    "    for edit in edit_pos:\n",
    "        o = sl[edit]\n",
    "        new_alpha = alpha - set([o])\n",
    "        new_char = list(new_alpha)[random.randint(0,2)]\n",
    "        sl[edit] = new_char\n",
    "        \n",
    "\n",
    "    s1 = gene\n",
    "    s2 = \"\".join(sl)\n",
    "    \n",
    "    return s1, s2\n",
    "    \n",
    "\n",
    "def generate_sequences(project_name, get_seq):\n",
    "    ![ -e $project_name/genome_data ] && rm -r $project_name/genome_data\n",
    "    !mkdir -p $project_name/genome_data\n",
    "    \n",
    "    s1, s2 = get_seq()\n",
    "    \n",
    "    records = []\n",
    "    record = SeqRecord(Seq(s1, DNAAlphabet), id='sample_gen_h1', description=\"hap 1\")    \n",
    "    records.append(record)\n",
    "    record = SeqRecord(Seq(s2, DNAAlphabet), id=\"sample_gen_h2\", description=\"hap 2\")    \n",
    "    records.append(record)\n",
    "    SeqIO.write(records, project_name + \"/genome_data/complete_genome.fa\", \"fasta\")\n",
    "    for rec in records:\n",
    "        SeqIO.write(rec, project_name + \"/genome_data/\" + rec.id + \".fasta\", \"fasta\")\n",
    "    \n",
    "    # record diff positions\n",
    "    file = open(project_name + '/genome_data/diff_pos.txt', 'w+')\n",
    "    for edit in range(len(s1)):\n",
    "        if s1[edit] != s2[edit]: file.write(str(edit + 1) + '\\n')\n",
    "    file.close()\n",
    "    \n",
    "    return s1, s2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Simulate Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_reads(project_name):\n",
    "    ![ -e $project_name/reads ] && rm -r $project_name/reads\n",
    "    !mkdir -p $project_name/reads\n",
    "    !cd $project_name/reads && simlord \\\n",
    "    --read-reference ../genome_data/complete_genome.fa \\\n",
    "    --coverage 50 \\\n",
    "    --probability-threshold 0.0001 \\\n",
    "    -pi 0.0001 \\\n",
    "    -pd 0.0001 \\\n",
    "    -ps 0.0001 \\\n",
    "    --fixed-readlength 25000 \\\n",
    "    --without-ns sd\n",
    "\n",
    "    !cat $project_name/reads/sd*.fastq > $project_name/reads/all.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Checking validity of SNP calling and haplotype phasing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. SNP calling and Phasing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def snp_and_phase(project_name, ref_hap_path):\n",
    "    ![ -e $project_name/mappings ] && rm -r $project_name/mappings\n",
    "    ![ -e $project_name/snps ] && rm -r $project_name/snps\n",
    "    ![ ! -e $project_name/mappings ] && mkdir -p $project_name/mappings \n",
    "    ![ ! -e $project_name/snps ] && mkdir -p $project_name/snps\n",
    "\n",
    "    reads_path = project_name + '/reads/all.fastq'\n",
    "    contigs_path = project_name + '/genome_data/sample_gen_h1.fasta'\n",
    "    contigs_path = ref_hap_path\n",
    "\n",
    "    # Indexing and Mapping\n",
    "    !minimap2 --cs -a $contigs_path $reads_path > $(pwd)/$project_name/mappings/mapping.sam\n",
    "    !samtools view -Sb $(pwd)/$project_name/mappings/mapping.sam > $(pwd)/$project_name/mappings/mapping.bam\n",
    "    !samtools sort $(pwd)/$project_name/mappings/mapping.bam > $(pwd)/$project_name/mappings/mapping.sorted.bam\n",
    "    !samtools index $(pwd)/$project_name/mappings/mapping.sorted.bam\n",
    "\n",
    "    # Phasing Methods\n",
    "    !samtools mpileup -q 20 -Q 0 $project_name/mappings/mapping.sorted.bam | python /media/anuradhawick/data/Tools/anaconda/anaconda3/lib/python3.6/site-packages/whatshap/simple_snp_caller.py --minabs 3 --minrel 0.25 --sample sample $contigs_path | awk '($1~/#/) || ($4~/[ACGT]/ && $5~/[ACGT]/)' > $(pwd)/$project_name/snps/initial_call.vcf\n",
    "    !whatshap genotype --ignore-read-groups --reference $contigs_path -o $(pwd)/$project_name/snps/genotyped.vcf $(pwd)/$project_name/snps/initial_call.vcf $project_name/mappings/mapping.sorted.bam\n",
    "    !whatshap phase --ignore-read-groups --reference $contigs_path -o $(pwd)/$project_name/snps/phased.vcf $(pwd)/$project_name/snps/genotyped.vcf $(pwd)/$project_name/mappings/mapping.sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Validating SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VCF parser\n",
    "import vcf\n",
    "\n",
    "def validate_snps(project_name):\n",
    "    snp_truth_path = project_name + '/genome_data/diff_pos.txt'\n",
    "    vcf_call_path = project_name + '/snps/initial_call.vcf'\n",
    "    with open(snp_truth_path) as f:\n",
    "        snp_positions = list(map(int, f.read().strip().split()))\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(vcf_call_path))\n",
    "    called_pos = [int(x.POS) for x in vcf_reader]\n",
    "\n",
    "    true_positives = set(called_pos).intersection(set(snp_positions))\n",
    "    false_negatives = set(snp_positions) - set(called_pos)\n",
    "    false_positives = set(called_pos) - set(snp_positions)\n",
    "\n",
    "    print(\"INFO::TRUE  POS: {0}\".format(len(true_positives)))\n",
    "    print(\"INFO::FALSE POS: {0}\".format(len(false_positives)))\n",
    "    print(\"INFO::FALSE NEG: {0}\".format(len(false_negatives)))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Validating Phasing Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_hamming_dist(s1, s2):\n",
    "    d = 0\n",
    "    for x in range(len(s1)):\n",
    "        if s1[x] != s2[x]:\n",
    "            d += 1\n",
    "    return d\n",
    "\n",
    "def validate_phasing(project_name, hap_1, hap_2):\n",
    "    phased_vcf_path = project_name + '/snps/phased.vcf'\n",
    "    \n",
    "    phased_dict = {}\n",
    "    unphased_dict = {}\n",
    "    phase_set_margins = {}\n",
    "    \n",
    "    for record in vcf.Reader(open(phased_vcf_path)):\n",
    "        pos = record.POS\n",
    "\n",
    "        # No haplotype information\n",
    "        if not record.samples[0].phased:\n",
    "            unphased_dict[pos - 1] = (record.REF, record.ALT)\n",
    "        else:\n",
    "            # phased\n",
    "            phase_set = record.samples[0].data.PS\n",
    "            phase_set_margins[phase_set] = pos\n",
    "\n",
    "            a, b = record.samples[0].data.GT.split(\"|\")\n",
    "            if int(a) == 0:\n",
    "                c1 = str(record.REF)\n",
    "                c2 = str(record.ALT[0])\n",
    "            else:\n",
    "                c2 = str(record.REF)\n",
    "                c1 = str(record.ALT[0])\n",
    "            phased_dict[pos - 1] = (phase_set, c1, c2)\n",
    "\n",
    "    original_haplotype_lst = list(hap_1)\n",
    "\n",
    "    phase_margin_indexes = sorted(list(phase_set_margins.values()))\n",
    "    phased_contigs = [] # array of tuples\n",
    "\n",
    "    hap1 = list(original_haplotype_lst)\n",
    "    hap2 = list(original_haplotype_lst)\n",
    "\n",
    "    for pos, var in phased_dict.items():\n",
    "        hap1[pos] = var[1]\n",
    "        hap2[pos] = var[2]\n",
    "\n",
    "    temp = 0\n",
    "    for pos in phase_margin_indexes:\n",
    "        hx1 = hap1[temp:pos]\n",
    "        hx2 = hap2[temp:pos]\n",
    "        phased_contigs.append((temp, pos, hx1, hx2))\n",
    "        temp = pos\n",
    "\n",
    "    for s, e, h1, h2 in  phased_contigs:\n",
    "        seq1_part = hap_1[s:e]\n",
    "        seq2_part = hap_2[s:e]\n",
    "        hap1_str = ''.join(h1)\n",
    "        hap2_str = ''.join(h2)\n",
    "\n",
    "        print('INFO::Hamming distances b/w s1/h1 = {0}'.format(get_hamming_dist(seq1_part, hap1_str)))\n",
    "        print('INFO::Hamming distances b/w s1/h2 = {0}'.format(get_hamming_dist(seq1_part, hap2_str)))\n",
    "        print('INFO::Hamming distances b/w s2/h1 = {0}'.format(get_hamming_dist(seq2_part, hap1_str)))\n",
    "        print('INFO::Hamming distances b/w s2/h2 = {0}'.format(get_hamming_dist(seq2_part, hap2_str)))\n",
    "\n",
    "        print(\"=======================================\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment Exec 1MB Het 1%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for reading/generating the reference: 0:00:00.071533 h\n",
      "Time for simulation of 4000 reads: 0:01:06.016984 h.\n",
      "[M::mm_idx_gen::0.023*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.030*1.48] sorted minimizers\n",
      "[M::main::0.031*1.48] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.033*1.44] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.035*1.42] distinct minimizers: 186691 (99.94% are singletons); average occurrences: 1.001; average spacing: 5.353\n",
      "[M::worker_pipeline::7.818*2.82] mapped 4000 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --cs -a simulated_haps_1mb/genome_data/sample_gen_h1.fasta simulated_haps_1mb/reads/all.fastq\n",
      "[M::main] Real time: 7.822 sec; CPU: 22.072 sec; Peak RSS: 0.204 GB\n",
      "[mpileup] 1 samples in 1 input files\n",
      "running only genotyping algorithm\n",
      "This is WhatsHap (genotyping) 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Initial genotyping of sample\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 4000 reads covering 9945 variants\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 4000 reads covering 9945 variants\n",
      "Kept 4000 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 468 reads covering 9929 variants\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 9929\n",
      "Genotype 1 sample by solving the genotyping problem ...\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.196 GB\n",
      "Time spent reading BAM:                         9.2 s\n",
      "Time spent parsing VCF:                         0.2 s\n",
      "Time spent selecting reads:                     1.8 s\n",
      "Time spent genotyping:                           105.0 s\n",
      "Time spent writing VCF:                         0.4 s\n",
      "Time spent on rest:                             0.0 s\n",
      "Total elapsed time:                           116.7 s\n",
      "This is WhatsHap 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Number of variants skipped due to missing genotypes: 0\n",
      "Number of remaining heterozygous variants: 9945\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 4000 reads covering 9945 variants\n",
      "Kept 4000 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 468 reads covering 9929 variants\n",
      "Best-case phasing would result in 1 non-singleton phased blocks (1 in total)\n",
      "... after read selection: 2 non-singleton phased blocks (2 in total)\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 9929\n",
      "Phasing 1 sample by solving the MEC problem ...\n",
      "MEC cost: 30\n",
      "No. of phased blocks: 2\n",
      "Largest component contains 5928 variants (59.7% of accessible variants) between position 412337 and 996206\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.213 GB\n",
      "Time spent reading BAM/CRAM:                    4.6 s\n",
      "Time spent parsing VCF:                         0.3 s\n",
      "Time spent selecting reads:                     1.6 s\n",
      "Time spent phasing:                            19.6 s\n",
      "Time spent writing VCF:                         0.5 s\n",
      "Time spent finding components:                  0.2 s\n",
      "Time spent on rest:                             2.3 s\n",
      "Total elapsed time:                            29.0 s\n",
      "INFO::TRUE  POS: 9945\n",
      "INFO::FALSE POS: 0\n",
      "INFO::FALSE NEG: 55\n",
      "\n",
      "INFO::Hamming distances b/w s1/h1 = 0\n",
      "INFO::Hamming distances b/w s1/h2 = 4001\n",
      "INFO::Hamming distances b/w s2/h1 = 4032\n",
      "INFO::Hamming distances b/w s2/h2 = 31\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 0\n",
      "INFO::Hamming distances b/w s1/h2 = 5928\n",
      "INFO::Hamming distances b/w s2/h1 = 5944\n",
      "INFO::Hamming distances b/w s2/h2 = 16\n",
      "=======================================\n"
     ]
    }
   ],
   "source": [
    "project_name = 'simulated_haps_1mb_h1'\n",
    "\n",
    "def get_seq():\n",
    "    return get_haps(0.01, 1000)\n",
    "\n",
    "hap_1, hap_2 = generate_sequences(project_name, get_seq)\n",
    "\n",
    "generate_reads(project_name)\n",
    "ref_hap_path = project_name + '/genome_data/sample_gen_h1.fasta'\n",
    "snp_and_phase(project_name)\n",
    "validate_snps(project_name)\n",
    "validate_phasing(project_name, hap_1, hap_2)\n",
    "\n",
    "project_name = 'simulated_haps_1mb_h1'\n",
    "ref_hap_path = project_name + '/genome_data/sample_gen_h1.fasta'\n",
    "snp_and_phase(project_name)\n",
    "validate_snps(project_name)\n",
    "validate_phasing(project_name, hap_1, hap_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment Exec 0.2,0.2,0.2,0.2,0.2 KB alternate regions Het 1%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for reading/generating the reference: 0:00:00.072570 h\n",
      "Time for simulation of 4000 reads: 0:01:07.011181 h.\n",
      "[M::mm_idx_gen::0.025*1.03] collected minimizers\n",
      "[M::mm_idx_gen::0.034*1.51] sorted minimizers\n",
      "[M::main::0.034*1.51] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.036*1.47] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.038*1.45] distinct minimizers: 186390 (99.93% are singletons); average occurrences: 1.001; average spacing: 5.361\n",
      "[M::worker_pipeline::8.171*2.82] mapped 4000 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --cs -a simulated_haps_0.2_0.2_0.2_0.2_0.2_alternate/genome_data/sample_gen_h1.fasta simulated_haps_0.2_0.2_0.2_0.2_0.2_alternate/reads/all.fastq\n",
      "[M::main] Real time: 8.174 sec; CPU: 23.039 sec; Peak RSS: 0.204 GB\n",
      "[mpileup] 1 samples in 1 input files\n",
      "running only genotyping algorithm\n",
      "This is WhatsHap (genotyping) 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Initial genotyping of sample\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 2624 reads covering 5992 variants\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 2624 reads covering 5992 variants\n",
      "Kept 2622 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 328 reads covering 5912 variants\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 5912\n",
      "Genotype 1 sample by solving the genotyping problem ...\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.161 GB\n",
      "Time spent reading BAM:                         6.5 s\n",
      "Time spent parsing VCF:                         0.1 s\n",
      "Time spent selecting reads:                     1.2 s\n",
      "Time spent genotyping:                            64.6 s\n",
      "Time spent writing VCF:                         0.3 s\n",
      "Time spent on rest:                             0.0 s\n",
      "Total elapsed time:                            72.6 s\n",
      "This is WhatsHap 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Number of variants skipped due to missing genotypes: 0\n",
      "Number of remaining heterozygous variants: 5981\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 2624 reads covering 5981 variants\n",
      "Kept 2622 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 328 reads covering 5914 variants\n",
      "Best-case phasing would result in 3 non-singleton phased blocks (3 in total)\n",
      "... after read selection: 5 non-singleton phased blocks (5 in total)\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 5914\n",
      "Phasing 1 sample by solving the MEC problem ...\n",
      "MEC cost: 30\n",
      "No. of phased blocks: 5\n",
      "Largest component contains 2000 variants (33.8% of accessible variants) between position 400044 and 599905\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.170 GB\n",
      "Time spent reading BAM/CRAM:                    2.9 s\n",
      "Time spent parsing VCF:                         0.2 s\n",
      "Time spent selecting reads:                     1.2 s\n",
      "Time spent phasing:                            13.0 s\n",
      "Time spent writing VCF:                         0.3 s\n",
      "Time spent finding components:                  0.1 s\n",
      "Time spent on rest:                             1.5 s\n",
      "Total elapsed time:                            19.2 s\n",
      "INFO::TRUE  POS: 5992\n",
      "INFO::FALSE POS: 0\n",
      "INFO::FALSE NEG: 8\n",
      "\n",
      "INFO::Hamming distances b/w s1/h1 = 690\n",
      "INFO::Hamming distances b/w s1/h2 = 0\n",
      "INFO::Hamming distances b/w s2/h1 = 17\n",
      "INFO::Hamming distances b/w s2/h2 = 707\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 0\n",
      "INFO::Hamming distances b/w s1/h2 = 1264\n",
      "INFO::Hamming distances b/w s2/h1 = 1293\n",
      "INFO::Hamming distances b/w s2/h2 = 29\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 0\n",
      "INFO::Hamming distances b/w s1/h2 = 2000\n",
      "INFO::Hamming distances b/w s2/h1 = 2000\n",
      "INFO::Hamming distances b/w s2/h2 = 0\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 0\n",
      "INFO::Hamming distances b/w s1/h2 = 411\n",
      "INFO::Hamming distances b/w s2/h1 = 411\n",
      "INFO::Hamming distances b/w s2/h2 = 0\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 1549\n",
      "INFO::Hamming distances b/w s1/h2 = 0\n",
      "INFO::Hamming distances b/w s2/h1 = 38\n",
      "INFO::Hamming distances b/w s2/h2 = 1587\n",
      "=======================================\n"
     ]
    }
   ],
   "source": [
    "project_name = 'simulated_haps_0.2_0.2_0.2_0.2_0.2_alternate'\n",
    "\n",
    "def get_seq():\n",
    "    s11, s12 = get_haps(0.01, 200)\n",
    "    s21 = s22 = createFrag(200)\n",
    "    s31, s32 = get_haps(0.01, 200)\n",
    "    s41 = s42 = createFrag(200)\n",
    "    s51, s52 = get_haps(0.01, 200)\n",
    "    \n",
    "    return s11 + s21 + s31 + s41 + s51, s12 + s22 + s32 + s42 + s52\n",
    "\n",
    "hap_1, hap_2 = generate_sequences(project_name, get_seq)\n",
    "generate_reads(project_name)\n",
    "ref_hap_path = project_name + '/genome_data/sample_gen_h1.fasta'\n",
    "snp_and_phase(project_name)\n",
    "validate_snps(project_name)\n",
    "validate_phasing(project_name, hap_1, hap_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for reading/generating the reference: 0:00:00.075196 h\n",
      "Time for simulation of 4000 reads: 0:01:07.760454 h.\n",
      "[M::mm_idx_gen::0.025*1.03] collected minimizers\n",
      "[M::mm_idx_gen::0.033*1.47] sorted minimizers\n",
      "[M::main::0.033*1.47] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.035*1.44] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.037*1.42] distinct minimizers: 186784 (99.92% are singletons); average occurrences: 1.001; average spacing: 5.350\n",
      "[M::worker_pipeline::7.817*2.81] mapped 4000 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --cs -a simulated_haps_0.2_0.6_0.2_alternate/genome_data/sample_gen_h1.fasta simulated_haps_0.2_0.6_0.2_alternate/reads/all.fastq\n",
      "[M::main] Real time: 7.828 sec; CPU: 21.971 sec; Peak RSS: 0.203 GB\n",
      "[mpileup] 1 samples in 1 input files\n",
      "running only genotyping algorithm\n",
      "This is WhatsHap (genotyping) 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Initial genotyping of sample\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 1626 reads covering 3980 variants\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 1626 reads covering 3980 variants\n",
      "Kept 1624 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 214 reads covering 3980 variants\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 3980\n",
      "Genotype 1 sample by solving the genotyping problem ...\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.138 GB\n",
      "Time spent reading BAM:                         4.0 s\n",
      "Time spent parsing VCF:                         0.1 s\n",
      "Time spent selecting reads:                     0.8 s\n",
      "Time spent genotyping:                            48.2 s\n",
      "Time spent writing VCF:                         0.2 s\n",
      "Time spent on rest:                             0.0 s\n",
      "Total elapsed time:                            53.2 s\n",
      "This is WhatsHap 0.18 running under Python 3.6.8\n",
      "Working on 1 samples from 1 family\n",
      "======== Working on chromosome 'sample_gen_h1'\n",
      "---- Processing individual sample\n",
      "Using maximum coverage per sample of 15X\n",
      "Number of variants skipped due to missing genotypes: 0\n",
      "Number of remaining heterozygous variants: 3980\n",
      "Reading alignments and detecting alleles ...\n",
      "Found 1626 reads covering 3980 variants\n",
      "Kept 1624 reads that cover at least two variants each\n",
      "Reducing coverage to at most 15X by selecting most informative reads ...\n",
      "Selected 214 reads covering 3980 variants\n",
      "Best-case phasing would result in 2 non-singleton phased blocks (2 in total)\n",
      "... after read selection: 2 non-singleton phased blocks (2 in total)\n",
      "Variants covered by at least one phase-informative read in at least one individual after read selection: 3980\n",
      "Phasing 1 sample by solving the MEC problem ...\n",
      "MEC cost: 30\n",
      "No. of phased blocks: 2\n",
      "Largest component contains 1993 variants (50.1% of accessible variants) between position 800023 and 998822\n",
      "======== Writing VCF\n",
      "Done writing VCF\n",
      "\n",
      "== SUMMARY ==\n",
      "Maximum memory usage: 0.148 GB\n",
      "Time spent reading BAM/CRAM:                    2.0 s\n",
      "Time spent parsing VCF:                         0.1 s\n",
      "Time spent selecting reads:                     0.7 s\n",
      "Time spent phasing:                             9.5 s\n",
      "Time spent writing VCF:                         0.2 s\n",
      "Time spent finding components:                  0.1 s\n",
      "Time spent on rest:                             0.9 s\n",
      "Total elapsed time:                            13.6 s\n",
      "INFO::TRUE  POS: 3980\n",
      "INFO::FALSE POS: 0\n",
      "INFO::FALSE NEG: 20\n",
      "\n",
      "INFO::Hamming distances b/w s1/h1 = 1987\n",
      "INFO::Hamming distances b/w s1/h2 = 0\n",
      "INFO::Hamming distances b/w s2/h1 = 13\n",
      "INFO::Hamming distances b/w s2/h2 = 2000\n",
      "=======================================\n",
      "INFO::Hamming distances b/w s1/h1 = 1993\n",
      "INFO::Hamming distances b/w s1/h2 = 0\n",
      "INFO::Hamming distances b/w s2/h1 = 0\n",
      "INFO::Hamming distances b/w s2/h2 = 1993\n",
      "=======================================\n"
     ]
    }
   ],
   "source": [
    "project_name = 'simulated_haps_0.2_0.6_0.2_alternate'\n",
    "\n",
    "def get_seq():\n",
    "    s11, s12 = get_haps(0.01, 200)\n",
    "    s21 = s22 = createFrag(600)\n",
    "    s31, s32 = get_haps(0.01, 200)\n",
    "    \n",
    "    return s11 + s21 + s31, s12 + s22 + s32\n",
    "\n",
    "hap_1, hap_2 = generate_sequences(project_name, get_seq)\n",
    "generate_reads(project_name)\n",
    "ref_hap_path = project_name + '/genome_data/sample_gen_h1.fasta'\n",
    "snp_and_phase(project_name, ref_hap_path)\n",
    "validate_snps(project_name)\n",
    "validate_phasing(project_name, hap_1, hap_2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
