{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Two Haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import DNAAlphabet\n",
    "\n",
    "heterozygous_rate = 0.01\n",
    "\n",
    "alpha = set(['A', 'C', 'G', 'T'])\n",
    "records = []\n",
    "for seq_record in SeqIO.parse(\"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/all_chr.fa\", \"fasta\"):\n",
    "    s_len = len(seq_record)\n",
    "    s = seq_record.seq\n",
    "    sl = list(s)\n",
    "    delta = round(s_len * heterozygous_rate)\n",
    "    edit_pos = random.sample(range(0, s_len), delta)\n",
    "    \n",
    "    for edit in edit_pos:\n",
    "        o = sl[edit]\n",
    "        new_alpha = alpha - set([o])\n",
    "        new_char = list(new_alpha)[random.randint(0,2)]\n",
    "        sl[edit] = new_char\n",
    "    \n",
    "    record = SeqRecord(Seq(\"\".join(sl), DNAAlphabet), id=seq_record.id + \"_h2\", description=\"other hap\")    \n",
    "    records.append(record)\n",
    "\n",
    "SeqIO.write(records, \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/hap2.fa\", \"fasta\")\n",
    "    \n",
    "!cat \\\n",
    "/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/all_chr.fa \\\n",
    "/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/hap2.fa \\\n",
    "> /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/with_hap.fasta\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./haplotype_yeast/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa -t 16 /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/with_hap.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-22 09:44:25\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ==> YeastGenome\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/with_hap.fasta ==> with_hap\n",
      "\n",
      "2019-01-22 09:44:26\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    YeastGenome.fa, length = 12157105, num fragments = 1, GC % = 38.15\n",
      "  Contig files: \n",
      "    with_hap\n",
      "  Calculating N50 and L50...\n",
      "    with_hap, N50 = 924431, L50 = 11, Total length = 24314210, GC % = 38.23, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing with_hap GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/basic_stats/with_hap_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-22 09:44:28\n",
      "Running Contig analyzer...\n",
      "  with_hap\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/contigs_report_with_hap.stdout and contigs_report_with_hap.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-22 09:44:33\n",
      "Running NA-NGA calculation...\n",
      "  with_hap, Largest alignment = 1531933, NA50 = 924431, NGA50 = 1090940, LA50 = 11, LGA50 = 5\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-22 09:44:33\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  with_hap\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-22 09:44:36\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-22 09:44:37\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Haplotype_genome_assembly/haplotype_yeast/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-22 09:44:37\n",
      "Elapsed time: 0:00:12.266589\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "project_name = 'haplotype_yeast'\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa\" -t 16 \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/with_hap.fasta\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
