{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Improving the Graph\n",
    "\n",
    "## Steps\n",
    "\n",
    "* Get the contigs and calculate the coverage\n",
    "    * Align the contigs to reads using Minimap2\n",
    "    * Sort the aligments using Samtools\n",
    "    * Obtain the dept of each contig position using Samtools depth\n",
    "\n",
    "* Get the coverage of links in the graph\n",
    "    * For each link, obtain the connection sequence by connecting the latter 50% of first contig with first 50% of the next contig\n",
    "    * Obtain the coverage as in the same way above"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting Contigs, Links and Connection Contigs from the GFA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'contig_1': 624999, 'contig_2': 1999806, 'contig_3': 1749098}\n"
     ]
    }
   ],
   "source": [
    "![ ! -e 'links' ] && mkdir 'links'\n",
    "![ ! -e 'contigs' ] && mkdir 'contigs'\n",
    "![ ! -e 'mappings' ] && mkdir 'mappings'\n",
    "\n",
    "\n",
    "import gfapy as gf\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import DNAAlphabet\n",
    "from Bio import SeqIO\n",
    "\n",
    "gfa = gf.Gfa.from_file('./graphs/assembly_graph.gfa')\n",
    "\n",
    "segments = {}\n",
    "contig_lengths = {}\n",
    "\n",
    "\n",
    "def get_aligning_sequences(s1, o1, s2, o2):\n",
    "    seq1 = Seq(s1)\n",
    "    seq2 = Seq(s2)\n",
    "\n",
    "    if (o1 == '-'):\n",
    "        seq1 = seq1.reverse_complement()\n",
    "    \n",
    "    if (o2 == '-'):\n",
    "        seq2 = seq2.reverse_complement()\n",
    "\n",
    "    return seq1, seq2\n",
    "\n",
    "for x in gfa.segments:\n",
    "    segments[x.name] = x\n",
    "    contig_lengths[x.name] = len(x.sequence)\n",
    "    record1 = SeqRecord(Seq(x.sequence, DNAAlphabet),\n",
    "                        id=x.name, description=\"Contig name=\"+x.name)\n",
    "    SeqIO.write(record1, \"./contigs/\" + x.name, \"fasta\")\n",
    "    \n",
    "    \n",
    "\n",
    "for x in gfa.dovetails:\n",
    "    seq1, seq2 = get_aligning_sequences(x.from_segment.sequence, x.from_orient, x.to_segment.sequence, x.to_orient)\n",
    "\n",
    "    idstr = x.from_segment.name + x.from_orient + x.to_segment.name + x.to_orient\n",
    "#   TODO check if we need to get half or a weighted amount considering the read lengths\n",
    "    record = SeqRecord(seq1[round(-1*len(seq1)/2)::] + seq2[0:int(len(seq1)/2)],\n",
    "                        id=idstr, description=\"\")\n",
    "\n",
    "    SeqIO.write(record, \"./links/\"+idstr, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computing Coverage of Links and Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to Obtain Coverage of Contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "def compute_contig_coverage(folder):\n",
    "    paths = Path(folder).glob('**/*')\n",
    "    readspath = \"../Assembly/Sampled Reads/\"\n",
    "\n",
    "    coverage_dict = {}\n",
    "\n",
    "    for p in [str(p) for p in paths]:\n",
    "        fname = p.split(\"/\").pop()\n",
    "        print(\"INFO::Inspecting path = \" + p)\n",
    "\n",
    "        !minimap2 -a \"$p\" \"$readspath\"sd_0004.fastq.gz > out.txt\n",
    "        !samtools sort out.txt > out.sorted.txt\n",
    "        !samtools depth out.sorted.txt > out.txt\n",
    "\n",
    "        tot = !cut -f3 out.txt | paste -sd+ | bc\n",
    "        lines = !wc -l out.txt\n",
    "\n",
    "        tot = float(tot[0])\n",
    "        lines = float(lines[0].split()[0])\n",
    "\n",
    "        coverage = tot/lines\n",
    "        !rm out.txt\n",
    "        !rm out.sorted.txt\n",
    "\n",
    "        coverage_dict[fname] = coverage\n",
    "    return coverage_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO::Inspecting path = contigs/contig_3\n",
      "[M::mm_idx_gen::0.043*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.053*1.36] sorted minimizers\n",
      "[M::main::0.053*1.36] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.056*1.34] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.060*1.32] distinct minimizers: 325791 (99.89% are singletons); average occurrences: 1.001; average spacing: 5.363\n",
      "[M::worker_pipeline::10.621*2.35] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a contigs/contig_3 ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 10.640 sec; CPU: 24.987 sec; Peak RSS: 0.493 GB\n",
      "INFO::Inspecting path = contigs/contig_2\n",
      "[M::mm_idx_gen::0.072*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.083*1.29] sorted minimizers\n",
      "[M::main::0.083*1.29] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.088*1.27] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.092*1.26] distinct minimizers: 372344 (99.87% are singletons); average occurrences: 1.001; average spacing: 5.364\n",
      "[M::worker_pipeline::11.314*2.45] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a contigs/contig_2 ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 11.318 sec; CPU: 27.718 sec; Peak RSS: 0.503 GB\n",
      "INFO::Inspecting path = contigs/contig_1\n",
      "[M::mm_idx_gen::0.037*1.04] collected minimizers\n",
      "[M::mm_idx_gen::0.045*1.35] sorted minimizers\n",
      "[M::main::0.045*1.35] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.046*1.34] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.048*1.33] distinct minimizers: 116663 (99.96% are singletons); average occurrences: 1.000; average spacing: 5.355\n",
      "[M::worker_pipeline::8.962*2.32] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a contigs/contig_1 ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 8.965 sec; CPU: 20.810 sec; Peak RSS: 0.481 GB\n",
      "{'contig_3': 49.193007481570504, 'contig_2': 51.34579254187656, 'contig_1': 99.67630348208557}\n"
     ]
    }
   ],
   "source": [
    "# Running Coverage Calculation for Contigs\n",
    "\n",
    "contig_coverage = compute_coverage(\"./contigs/\")\n",
    "\n",
    "print(contig_coverage)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to Obtain Coverage of Links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def get_match_from_cigar(cigar):\n",
    "    matches = re.findall(r'(\\d+)M', cigar)\n",
    "    misses = re.findall(r'(\\d+)H', cigar)\n",
    "    deletions = re.findall(r'(\\d+)D', cigar)\n",
    "    insertions = re.findall(r'(\\d+)I', cigar)\n",
    "    \n",
    "    return sum(map(int, matches)), sum(map(int, misses)), sum(map(int, deletions)), sum(map(int, insertions))\n",
    "    \n",
    "\n",
    "def obtain_relevant_mappings(mapping_file, output_file, contig1_length, contig2_length):\n",
    "    with open(output_file, 'w+') as outf:\n",
    "        with open(mapping_file, 'r') as mapf:\n",
    "            line_o = mapf.readline()\n",
    "            line = line_o.split()\n",
    "            # consume headers\n",
    "            while line_o[0] == \"@\":\n",
    "                outf.write(line_o)\n",
    "                line_o = mapf.readline()\n",
    "                line = line_o.split()\n",
    "            while line_o:\n",
    "                pos = int(line[3])\n",
    "\n",
    "                if pos == 0: \n",
    "                    line_o = mapf.readline()\n",
    "                    line = line_o.split()\n",
    "                    continue # no match found\n",
    "                else:\n",
    "                    # match found\n",
    "                    matches, misses, deletions, insertions = get_match_from_cigar(line[5])\n",
    "                    \n",
    "                    # ensure mappings including both links are taken\n",
    "                    match_starts_contig1 = (pos-1) < contig1_length\n",
    "                    match_ends_contig2 = matches + pos > contig1_length\n",
    "\n",
    "                    # if matches are from both links it is a valid mapping\n",
    "                    if match_starts_contig1 and match_ends_contig2:# and matches / (matches + misses + deletions + insertions) > 0.75:\n",
    "                        outf.write(line_o)\n",
    "                \n",
    "                line_o = mapf.readline()\n",
    "                line = line_o.split()\n",
    "    \n",
    "def get_coverage_from_file_name(fname):\n",
    "    content = re.split('\\+|\\-',fname)\n",
    "    valid_content = []\n",
    "    \n",
    "    for x in content:\n",
    "        if len(x.strip()) > 0:\n",
    "            valid_content.append(x.strip())\n",
    "    \n",
    "    return contig_coverage[valid_content[0]], contig_coverage[valid_content[1]]\n",
    "    \n",
    "def compute_link_coverage(folder):\n",
    "    paths = Path(folder).glob('**/*')\n",
    "    readspath = \"../Assembly/Sampled Reads/\"\n",
    "\n",
    "    coverage_dict = {}\n",
    "\n",
    "    for p in [str(p) for p in paths]:\n",
    "        fname = p.split(\"/\").pop()\n",
    "        \n",
    "        c1_len, c2_len = get_coverage_from_file_name(fname)\n",
    "        \n",
    "        print(\"INFO::Inspecting link = \" + p)\n",
    "\n",
    "        !minimap2 -a \"$p\" \"$readspath\"sd_0004.fastq.gz > unfiltered.out.txt\n",
    "        \n",
    "        obtain_relevant_mappings('unfiltered.out.txt', 'filtered.out.txt', c1_len, c2_len)\n",
    "        \n",
    "        !samtools sort filtered.out.txt > filtered.out.sorted.txt\n",
    "        !samtools depth filtered.out.sorted.txt > out.txt\n",
    "\n",
    "        tot = !cut -f3 out.txt | paste -sd+ | bc\n",
    "        lines = !wc -l out.txt\n",
    "\n",
    "        tot = float(tot[0])\n",
    "        lines = float(lines[0].split()[0])\n",
    "\n",
    "        coverage = tot/lines\n",
    "        !rm out.txt\n",
    "        !rm unfiltered.out.txt\n",
    "        !rm filtered.out.txt\n",
    "        !rm filtered.out.sorted.txt\n",
    "\n",
    "        coverage_dict[fname] = coverage\n",
    "    return coverage_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO::Inspecting link = links/contig_1-contig_3-\n",
      "[M::mm_idx_gen::0.017*1.04] collected minimizers\n",
      "[M::mm_idx_gen::0.023*1.55] sorted minimizers\n",
      "[M::main::0.023*1.55] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.026*1.50] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.027*1.47] distinct minimizers: 116533 (99.96% are singletons); average occurrences: 1.000; average spacing: 5.361\n",
      "[M::worker_pipeline::7.603*2.21] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a links/contig_1-contig_3- ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 7.606 sec; CPU: 16.801 sec; Peak RSS: 0.490 GB\n",
      "DONE CORRECT MAPPING\n",
      "INFO::Inspecting link = links/contig_1+contig_1+\n",
      "[M::mm_idx_gen::0.017*1.04] collected minimizers\n",
      "[M::mm_idx_gen::0.023*1.53] sorted minimizers\n",
      "[M::main::0.023*1.53] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.025*1.49] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.027*1.46] distinct minimizers: 116659 (99.96% are singletons); average occurrences: 1.000; average spacing: 5.355\n",
      "[M::worker_pipeline::8.588*2.29] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a links/contig_1+contig_1+ ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 8.591 sec; CPU: 19.638 sec; Peak RSS: 0.495 GB\n",
      "DONE CORRECT MAPPING\n",
      "INFO::Inspecting link = links/contig_3+contig_2+\n",
      "[M::mm_idx_gen::0.044*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.054*1.36] sorted minimizers\n",
      "[M::main::0.054*1.36] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.058*1.33] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.062*1.31] distinct minimizers: 325581 (99.89% are singletons); average occurrences: 1.001; average spacing: 5.366\n",
      "[M::worker_pipeline::10.889*2.42] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a links/contig_3+contig_2+ ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 10.914 sec; CPU: 26.430 sec; Peak RSS: 0.508 GB\n",
      "DONE CORRECT MAPPING\n",
      "INFO::Inspecting link = links/contig_1+contig_2+\n",
      "[M::mm_idx_gen::0.016*1.05] collected minimizers\n",
      "[M::mm_idx_gen::0.021*1.55] sorted minimizers\n",
      "[M::main::0.021*1.55] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.023*1.51] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.024*1.49] distinct minimizers: 113384 (97.09% are singletons); average occurrences: 1.029; average spacing: 5.357\n",
      "[M::worker_pipeline::7.840*2.23] mapped 19248 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a links/contig_1+contig_2+ ../Assembly/Sampled Reads/sd_0004.fastq.gz\n",
      "[M::main] Real time: 7.842 sec; CPU: 17.484 sec; Peak RSS: 0.497 GB\n",
      "DONE CORRECT MAPPING\n",
      "{'contig_1-contig_3-': 26.783238336499895, 'contig_1+contig_1+': 31.54664988110225, 'contig_3+contig_2+': 14.922250671329545, 'contig_1+contig_2+': 31.54664988110225}\n"
     ]
    }
   ],
   "source": [
    "# Running Coverage Calculation for Contigs and Links\n",
    "\n",
    "link_coverage = compute_link_coverage(\"./links/\")\n",
    "\n",
    "print(link_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{   'contig_1': 99.67630348208557,\n",
      "    'contig_2': 51.34579254187656,\n",
      "    'contig_3': 49.193007481570504}\n",
      "\n",
      "{   'contig_1+contig_1+': 31.54664988110225,\n",
      "    'contig_1+contig_2+': 31.54664988110225,\n",
      "    'contig_1-contig_3-': 26.783238336499895,\n",
      "    'contig_3+contig_2+': 14.922250671329545}\n"
     ]
    }
   ],
   "source": [
    "import pprint\n",
    "\n",
    "pp = pprint.PrettyPrinter(indent=4)\n",
    "\n",
    "pp.pprint(contig_coverage)\n",
    "print()\n",
    "pp.pprint(link_coverage)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
