{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Improving the Graph\n",
    "\n",
    "## Steps\n",
    "\n",
    "* Get the contigs and calculate the coverage\n",
    "    * Align the contigs to reads using Minimap2\n",
    "    * Sort the aligments using Samtools\n",
    "    * Obtain the dept of each contig position using Samtools depth\n",
    "\n",
    "* Get the coverage of links in the graph\n",
    "    * For each link, obtain the connection sequence by connecting the latter 50% of first contig with first 50% of the next contig\n",
    "    * Obtain the coverage as in the same way above"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting Contigs, Links and Connection Contigs from the GFA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All the imports\n",
    "import gfapy as gf\n",
    "\n",
    "# Python imports\n",
    "from pathlib import Path\n",
    "import re\n",
    "import copy\n",
    "from uuid import uuid1\n",
    "from joblib import Parallel, delayed\n",
    "import multiprocessing\n",
    "\n",
    "# BioPython imports\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import DNAAlphabet\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Scientific imports\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# for generated genome G4\n",
    "# project_name = \"simG4\"\n",
    "# readspath = \"../Assembly/Sampled Reads/sd_0004.fastq\"\n",
    "# graphpath = \"/media/anuradhawick/data/Experiments/Assembly Graph/Assembly/FLYE/Out FLYE_4/assembly_graph.gfa\"\n",
    "\n",
    "# e coli\n",
    "# readspath = \"../Assembly/Sampled Reads EColi/sd_0001.fastq\"\n",
    "# graphpath = \"/media/anuradhawick/data/Experiments/Assembly Graph/Assembly/FLYE/Out EColi/assembly_graph.gfa\"\n",
    "\n",
    "# paths for human chromosome 22 (5MB)\n",
    "readspath = \"../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\"\n",
    "graphpath = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/FLYE/Out Chr22/assembly_graph.gfa\"\n",
    "project_name = \"chr22\"\n",
    "\n",
    "# paths for Yeast Genome\n",
    "# readspath = \"../Assembly/Sampled_Reads_Yeast/sd_0001.fastq\"\n",
    "# graphpath = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/FLYE/Out Yeast/assembly_graph_refined.gfa\"\n",
    "# project_name = \"yeast_refined\"\n",
    "\n",
    "read_coverage = 50\n",
    "read_length = 20000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "![ ! -e \"$project_name\" ] && mkdir \"$project_name\"\n",
    "\n",
    "![ ! -e \"$project_name\"/links ] && mkdir \"$project_name\"/links\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "\n",
    "gfa = gf.Gfa.from_file(graphpath)\n",
    "\n",
    "segments = {}\n",
    "contig_lengths = {}\n",
    "\n",
    "def get_aligning_sequences(s1, o1, s2, o2):\n",
    "    seq1 = Seq(s1)\n",
    "    seq2 = Seq(s2)\n",
    "\n",
    "    if (o1 == '-'):\n",
    "        seq1 = seq1.reverse_complement()\n",
    "    \n",
    "    if (o2 == '-'):\n",
    "        seq2 = seq2.reverse_complement()\n",
    "\n",
    "    return seq1, seq2\n",
    "\n",
    "for x in gfa.segments:\n",
    "    segments[x.name] = x\n",
    "    contig_lengths[x.name] = len(x.sequence)\n",
    "    record1 = SeqRecord(Seq(x.sequence, DNAAlphabet),\n",
    "                        id=x.name, description=\"Contig name=\"+x.name)\n",
    "    SeqIO.write(record1, project_name + \"/contigs/\" + x.name, \"fasta\")\n",
    "    \n",
    "    \n",
    "\n",
    "for x in gfa.dovetails:\n",
    "    seq1, seq2 = get_aligning_sequences(x.from_segment.sequence, x.from_orient, x.to_segment.sequence, x.to_orient)\n",
    "\n",
    "    idstr = x.from_segment.name + x.from_orient + x.to_segment.name + x.to_orient\n",
    "#   TODO check if we need to get half or a weighted amount considering the read lengths\n",
    "    seq1_portion = round(read_length / 2)\n",
    "    seq2_portion = round(read_length / 2)\n",
    "    \n",
    "    record = SeqRecord(seq1[-1*seq1_portion::] + seq2[0:seq2_portion],\n",
    "                        id=idstr, description=\"\")\n",
    "\n",
    "    SeqIO.write(record, project_name + \"/links/\"+idstr, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computing Coverage of Links and Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to Obtain Coverage of Contigs and Links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_contig_coverage(folder):\n",
    "    paths = Path(folder).glob('**/*')\n",
    "\n",
    "    coverage_dict = {}\n",
    "    \n",
    "    for p in [str(p) for p in paths]:\n",
    "        print(\"INFO::Inspecting path = \" + p)\n",
    "        fname = p.split(\"/\").pop()\n",
    "                \n",
    "        !minimap2 --secondary=no -a \"$p\" \"$readspath\" > out.txt\n",
    "        !samtools sort out.txt > out.sorted.txt\n",
    "        !samtools depth out.sorted.txt > out.txt\n",
    "\n",
    "        tot = !cut -f3 out.txt | paste -sd+ | bc\n",
    "        lines = !wc -l out.txt\n",
    "\n",
    "        tot = float(tot[0])\n",
    "        lines = float(lines[0].split()[0])\n",
    "        \n",
    "#         y = !cut -f3 out.txt        \n",
    "        \n",
    "#         y = np.array(y)\n",
    "#         x = np.arange(len(y))\n",
    "        \n",
    "#         plt.plot(x,y)\n",
    "#         plt.show()\n",
    "#         plt.figure()\n",
    "\n",
    "        coverage = tot/lines\n",
    "        !rm out.txt\n",
    "        !rm out.sorted.txt\n",
    "\n",
    "        coverage_dict[fname] = coverage\n",
    "        \n",
    "        print(\"Coverage: \" + str(coverage))\n",
    "        \n",
    "    return coverage_dict\n",
    "\n",
    "\n",
    "def get_match_from_cigar(cigar):\n",
    "    matches = re.findall(r'(\\d+)M', cigar)\n",
    "    misses = re.findall(r'(\\d+)H', cigar)\n",
    "    deletions = re.findall(r'(\\d+)D', cigar)\n",
    "    insertions = re.findall(r'(\\d+)I', cigar)\n",
    "    \n",
    "    return sum(map(int, matches)), sum(map(int, misses)), sum(map(int, deletions)), sum(map(int, insertions))\n",
    "    \n",
    "\n",
    "def obtain_relevant_mappings(mapping_file, output_file, contig1_length, contig2_length):\n",
    "    print (\"INFO:: Filtering irrelavant maps\")\n",
    "    with open(output_file, 'w+') as outf:\n",
    "        with open(mapping_file, 'r') as mapf:\n",
    "            line_o = mapf.readline()\n",
    "            line = line_o.split()\n",
    "            # consume headers\n",
    "            while line_o[0] == \"@\":\n",
    "                outf.write(line_o)\n",
    "                line_o = mapf.readline()\n",
    "                line = line_o.split()\n",
    "            while line_o:\n",
    "                # mapping attributes\n",
    "                pos = int(line[3])\n",
    "\n",
    "                if pos == 0: \n",
    "                    line_o = mapf.readline()\n",
    "                    line = line_o.split()\n",
    "                    continue # no match found\n",
    "                else:\n",
    "                    # match found\n",
    "                    matches, misses, deletions, insertions = get_match_from_cigar(line[5])\n",
    "                    \n",
    "                    # ensure mappings including both links are taken (half of a read lenth overlap)\n",
    "                    match_starts_contig1 = (pos-1) < contig1_length * 0.75\n",
    "                    match_ends_contig2 = sum([matches, misses, deletions, insertions]) + pos > contig1_length + contig2_length * 0.25\n",
    "\n",
    "                    # if matches are from both links it is a valid mapping\n",
    "                    if match_starts_contig1 and match_ends_contig2: # and matches / (matches + misses + deletions + insertions) > 0.9:\n",
    "                        outf.write(line_o)\n",
    "                \n",
    "                line_o = mapf.readline()\n",
    "                line = line_o.split()\n",
    "    print (\"INFO:: Done filtering irrelavant maps\")\n",
    "    \n",
    "    \n",
    "def get_contig_names_from_file_name(fname):\n",
    "    content = re.split('\\+|\\-',fname)\n",
    "    valid_content = []\n",
    "    \n",
    "    for x in content:\n",
    "        if len(x.strip()) > 0:\n",
    "            valid_content.append(x.strip())\n",
    "    \n",
    "    return valid_content\n",
    "    \n",
    "    \n",
    "def compute_link_coverage(folder, read_coverage=50):\n",
    "    paths = Path(folder).glob('**/*')\n",
    "\n",
    "    coverage_dict = {}\n",
    "    \n",
    "    for p in [str(p) for p in paths]:\n",
    "        c1_len = round(read_length / 2)\n",
    "        c2_len = round(read_length / 2)   \n",
    "        fname = p.split(\"/\").pop()\n",
    "        \n",
    "        print(\"INFO::Inspecting link = \" + p)\n",
    "\n",
    "        !minimap2 --secondary=no -a \"$p\" \"$readspath\" > unfiltered.out.txt\n",
    "        \n",
    "        obtain_relevant_mappings('unfiltered.out.txt', 'filtered.out.txt', c1_len, c2_len)\n",
    "                \n",
    "        !samtools sort filtered.out.txt > filtered.out.sorted.txt\n",
    "        \n",
    "        basewise_coverage = !samtools depth filtered.out.sorted.txt\n",
    "        \n",
    "        data = [int(x.split()[2]) for x in basewise_coverage]\n",
    "        npd = np.array(data)\n",
    "\n",
    "#         n, bins, patches = plt.hist(x=npd, bins='auto', color='#0504aa',\n",
    "#                             alpha=0.7, rwidth=0.85)\n",
    "#         plt.grid(axis='y', alpha=0.75)\n",
    "#         plt.xlabel('Coverage')\n",
    "#         plt.ylabel('Occurances')\n",
    "#         plt.title('Coverage Distribution')\n",
    "#         maxfreq = n.max()\n",
    "#         # Set a clean upper y-axis limit.\n",
    "#         plt.ylim(ymax=np.ceil(maxfreq / 10) * 10 if maxfreq % 10 else maxfreq + 10)\n",
    "#         plt.show()\n",
    "#         plt.figure()\n",
    "        if len(npd) > 0:\n",
    "            mean = npd.mean()# because we consider half's coverage for the whole contigs (correction)\n",
    "            std = npd.std()\n",
    "        else:\n",
    "            mean, std = 0, 0\n",
    "        \n",
    "        print(\"Mean\", mean, \"Std\", std)\n",
    "\n",
    "        !rm unfiltered.out.txt\n",
    "        !rm filtered.out.txt\n",
    "        !rm filtered.out.sorted.txt\n",
    "\n",
    "        coverage_dict[fname] = mean\n",
    "        \n",
    "    return coverage_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining Coverage of Links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO::Inspecting path = chr22/contigs/contig_5\n",
      "[M::mm_idx_gen::0.001*1.49] collected minimizers\n",
      "[M::mm_idx_gen::0.001*2.06] sorted minimizers\n",
      "[M::main::0.001*2.04] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.002*1.99] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.002*1.95] distinct minimizers: 1290 (99.46% are singletons); average occurrences: 1.005; average spacing: 5.267\n",
      "[M::worker_pipeline::41.487*2.88] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/contigs/contig_5 ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 41.487 sec; CPU: 119.295 sec; Peak RSS: 0.530 GB\n",
      "Coverage: 988.8228663446055\n",
      "INFO::Inspecting path = chr22/contigs/contig_3\n",
      "[M::mm_idx_gen::0.019*0.65] collected minimizers\n",
      "[M::mm_idx_gen::0.035*0.60] sorted minimizers\n",
      "[M::main::0.035*0.60] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.036*0.61] mid_occ = 58\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.036*0.62] distinct minimizers: 48797 (97.37% are singletons); average occurrences: 1.095; average spacing: 5.295\n",
      "[M::worker_pipeline::214.664*2.91] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/contigs/contig_3 ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 214.666 sec; CPU: 624.178 sec; Peak RSS: 1.687 GB\n",
      "Coverage: 141.2423477772241\n",
      "INFO::Inspecting path = chr22/contigs/contig_2\n",
      "[M::mm_idx_gen::0.014*0.83] collected minimizers\n",
      "[M::mm_idx_gen::0.028*0.70] sorted minimizers\n",
      "[M::main::0.028*0.70] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.029*0.71] mid_occ = 39\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.030*0.72] distinct minimizers: 45547 (95.33% are singletons); average occurrences: 1.105; average spacing: 5.292\n",
      "[M::worker_pipeline::207.210*2.87] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/contigs/contig_2 ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 207.214 sec; CPU: 595.346 sec; Peak RSS: 1.940 GB\n",
      "Coverage: 155.69038673037727\n",
      "INFO::Inspecting path = chr22/contigs/contig_1\n",
      "[M::mm_idx_gen::0.211*0.80] collected minimizers\n",
      "[M::mm_idx_gen::0.278*0.85] sorted minimizers\n",
      "[M::main::0.278*0.85] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.289*0.86] mid_occ = 91\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.298*0.86] distinct minimizers: 715082 (96.27% are singletons); average occurrences: 1.153; average spacing: 5.313\n",
      "[M::worker_pipeline::65.670*2.79] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/contigs/contig_1 ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 65.685 sec; CPU: 183.507 sec; Peak RSS: 0.712 GB\n",
      "Coverage: 51.04089979036749\n",
      "INFO::Inspecting path = chr22/contigs/contig_4\n",
      "[M::mm_idx_gen::0.010*0.31] collected minimizers\n",
      "[M::mm_idx_gen::0.076*0.09] sorted minimizers\n",
      "[M::main::0.076*0.09] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.077*0.09] mid_occ = 5\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.077*0.10] distinct minimizers: 5483 (99.16% are singletons); average occurrences: 1.016; average spacing: 5.294\n",
      "[M::worker_pipeline::104.491*2.88] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/contigs/contig_4 ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 104.491 sec; CPU: 301.110 sec; Peak RSS: 0.564 GB\n",
      "Coverage: 491.391995930134\n",
      "INFO::Inspecting link = chr22/links/contig_2-contig_4-\n",
      "[M::mm_idx_gen::0.019*0.12] collected minimizers\n",
      "[M::mm_idx_gen::0.019*0.21] sorted minimizers\n",
      "[M::main::0.019*0.21] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.020*0.22] mid_occ = 20\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.020*0.22] distinct minimizers: 3719 (99.03% are singletons); average occurrences: 1.017; average spacing: 5.288\n",
      "[M::worker_pipeline::91.759*2.87] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/links/contig_2-contig_4- ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 91.759 sec; CPU: 263.120 sec; Peak RSS: 0.992 GB\n",
      "INFO:: Filtering irrelavant maps\n",
      "INFO:: Done filtering irrelavant maps\n",
      "Mean 562.0861369289174 Std 1811.9381087169909\n",
      "INFO::Inspecting link = chr22/links/contig_1+contig_4+\n",
      "[M::mm_idx_gen::0.008*0.24] collected minimizers\n",
      "[M::mm_idx_gen::0.016*0.21] sorted minimizers\n",
      "[M::main::0.016*0.21] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.017*0.22] mid_occ = 8\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.017*0.23] distinct minimizers: 3763 (99.42% are singletons); average occurrences: 1.010; average spacing: 5.263\n",
      "[M::worker_pipeline::88.271*2.85] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/links/contig_1+contig_4+ ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 88.272 sec; CPU: 251.653 sec; Peak RSS: 0.536 GB\n",
      "INFO:: Filtering irrelavant maps\n",
      "INFO:: Done filtering irrelavant maps\n",
      "Mean 126.58327427356485 Std 417.47387070105833\n",
      "INFO::Inspecting link = chr22/links/contig_2+contig_4+\n",
      "[M::mm_idx_gen::0.041*0.05] collected minimizers\n",
      "[M::mm_idx_gen::0.049*0.07] sorted minimizers\n",
      "[M::main::0.049*0.07] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.049*0.08] mid_occ = 10\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.049*0.08] distinct minimizers: 3624 (97.85% are singletons); average occurrences: 1.041; average spacing: 5.301\n",
      "[M::worker_pipeline::155.837*2.93] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/links/contig_2+contig_4+ ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 155.838 sec; CPU: 456.452 sec; Peak RSS: 1.577 GB\n",
      "INFO:: Filtering irrelavant maps\n",
      "INFO:: Done filtering irrelavant maps\n",
      "Mean 2454.0680703778485 Std 2442.7432330134598\n",
      "INFO::Inspecting link = chr22/links/contig_3+contig_4-\n",
      "[M::mm_idx_gen::0.039*0.06] collected minimizers\n",
      "[M::mm_idx_gen::0.051*0.07] sorted minimizers\n",
      "[M::main::0.051*0.08] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.051*0.08] mid_occ = 20\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.051*0.08] distinct minimizers: 3618 (96.05% are singletons); average occurrences: 1.044; average spacing: 5.292\n",
      "[M::worker_pipeline::64.390*2.88] mapped 10791 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 --secondary=no -a chr22/links/contig_3+contig_4- ../Assembly/Sampled_Reads_Chr22/sd_0001.fastq\n",
      "[M::main] Real time: 64.390 sec; CPU: 185.421 sec; Peak RSS: 0.548 GB\n",
      "INFO:: Filtering irrelavant maps\n",
      "INFO:: Done filtering irrelavant maps\n",
      "Mean 410.83259766056057 Std 1444.2069386494434\n",
      "{'contig_2-contig_4-': 562.0861369289174, 'contig_1+contig_4+': 126.58327427356485, 'contig_2+contig_4+': 2454.0680703778485, 'contig_3+contig_4-': 410.83259766056057}\n",
      "\n",
      "{'contig_5': 988.8228663446055, 'contig_3': 141.2423477772241, 'contig_2': 155.69038673037727, 'contig_1': 51.04089979036749, 'contig_4': 491.391995930134}\n",
      "Link coverages after removal of insignificant links\n",
      "{'contig_2-contig_4-': 562.0861369289174, 'contig_1+contig_4+': 126.58327427356485, 'contig_2+contig_4+': 2454.0680703778485, 'contig_3+contig_4-': 410.83259766056057}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Running Coverage Calculation for Contigs\n",
    "\n",
    "\n",
    "contig_coverage = compute_contig_coverage(project_name+\"/contigs/\")\n",
    "# # contig_coverage = {   'c2': 50,\n",
    "# #     'c1': 150,\n",
    "# #     'c3': 49,\n",
    "# #     'c4': 50\n",
    "# #     }\n",
    "\n",
    "link_coverage = compute_link_coverage(project_name+\"/links/\")\n",
    "# link_coverage = {   'c2+c1+': 50,\n",
    "#     'c1+c1+': 100,\n",
    "#     'c1+c3+': 53,\n",
    "#     'c3+c4': 50}\n",
    "\n",
    "print(link_coverage)\n",
    "print()\n",
    "print(contig_coverage)\n",
    "\n",
    "for key in list(link_coverage.keys()):\n",
    "    if link_coverage[key] == 0:\n",
    "        del link_coverage[key]\n",
    "        \n",
    "print(\"Link coverages after removal of insignificant links\")\n",
    "print(link_coverage)\n",
    "print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
