{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QUAST Assessment for FLYE Chr 22 5mb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "# Biopython imports\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import DNAAlphabet\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all -o ./chr22/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa -t 16 ./chr22/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-02 09:42:55\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa ==> chr22_5mb\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./chr22/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2019-01-02 09:42:56\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    chr22_5mb.fa, length = 5000000, num fragments = 1, GC % = 45.41\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 4381954, L50 = 1, Total length = 4967640, GC % = 45.64, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:42:57\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:43:01\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 4377749, NA50 = 4377749, NGA50 = 4377749, LA50 = 1, LGA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:43:01\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-02 09:43:02\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-02 09:43:03\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-02 09:43:03\n",
      "Elapsed time: 0:00:07.601909\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.117*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.140*1.33] sorted minimizers\n",
      "[M::main::0.140*1.33] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.150*1.31] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.159*1.29] distinct minimizers: 799170 (95.15% are singletons); average occurrences: 1.182; average spacing: 5.292\n",
      "[M::worker_pipeline::6.196*1.04] mapped 5 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa ./chr22/contigs/contigs.fasta\n",
      "[M::main] Real time: 6.201 sec; CPU: 6.470 sec; Peak RSS: 0.179 GB\n"
     ]
    }
   ],
   "source": [
    "# obtain all contigs to one file\n",
    "project_name = \"chr22\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa\"\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_1' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_2' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_3' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_4' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_5' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.303*1.00] collected minimizers\n",
      "[M::mm_idx_gen::0.392*1.37] sorted minimizers\n",
      "[M::main::0.392*1.37] loaded/built the index for 1000 target sequence(s)\n",
      "[M::mm_mapopt_update::0.408*1.36] mid_occ = 206\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1000\n",
      "[M::mm_idx_stat::0.418*1.35] distinct minimizers: 799170 (0.43% are singletons); average occurrences: 2.359; average spacing: 5.302\n",
      "[M::worker_pipeline::12.922*1.05] mapped 5 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a chr22_chunk_align/chunks/all.fa ./chr22_chunk_align/contigs/contigs.fasta\n",
      "[M::main] Real time: 12.928 sec; CPU: 13.638 sec; Peak RSS: 0.305 GB\n"
     ]
    }
   ],
   "source": [
    "# chunk and align\n",
    "\n",
    "# obtain all contigs to one file\n",
    "project_name = \"chr22_chunk_align\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa\"\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "!mkdir \"$project_name/chunks\"\n",
    "\n",
    "chunks = []\n",
    "with open(reference_genome, 'r') as f:\n",
    "    content = f.read().strip().split('\\n')\n",
    "    content.pop(0)\n",
    "    \n",
    "    string = \"\".join(content)\n",
    "    index = 0\n",
    "    \n",
    "    while index < len(string):\n",
    "        chunks.append(string[index:index+10000].upper())\n",
    "        index += 5000\n",
    "\n",
    "chunk_index = 0\n",
    "records = []\n",
    "for c in chunks:\n",
    "    r = SeqRecord(Seq(c, DNAAlphabet), id=\"Chunk_\"+str(chunk_index), description=\"Generated Chunk\" + str(chunk_index) + \"-\" + str(chunk_index + 10000))\n",
    "    SeqIO.write(r, project_name+\"/chunks/\"+str(chunk_index) + \".fa\", \"fasta\")\n",
    "    chunk_index += 5000\n",
    "    records.append(r)\n",
    "\n",
    "SeqIO.write(records, project_name+\"/chunks/all.fa\", \"fasta\")\n",
    "\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_1' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_2' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_3' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_4' \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_5' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "# !quast.py -a all -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$project_name\"/chunks/all.fa ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# # !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# # !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QUAST Assessment for Canu Chr 22 5mb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all -o ./chr22canu/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa -t 16 /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Chr22/r1.contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-42-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2018-12-20 14:00:14\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa ==> chr22_5mb\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Chr22/r1.contigs.fasta ==> r1.contigs\n",
      "\n",
      "2018-12-20 14:00:14\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    chr22_5mb.fa, length = 5000000, num fragments = 1, GC % = 45.41\n",
      "  Contig files: \n",
      "    r1.contigs\n",
      "  Calculating N50 and L50...\n",
      "    r1.contigs, N50 = 4995129, L50 = 1, Total length = 5020595, GC % = 45.44, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing r1.contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/basic_stats/r1.contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 14:00:15\n",
      "Running Contig analyzer...\n",
      "  r1.contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/contigs_report_r1-contigs.stdout and contigs_report_r1-contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 14:00:20\n",
      "Running NA-NGA calculation...\n",
      "  r1.contigs, Largest alignment = 4995129, NA50 = 4995129, NGA50 = 4995129, LA50 = 1, LGA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 14:00:20\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  r1.contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2018-12-20 14:00:21\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2018-12-20 14:00:22\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22canu/quast_eval/quast.log\n",
      "\n",
      "Finished: 2018-12-20 14:00:22\n",
      "Elapsed time: 0:00:08.471711\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.109*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.132*1.35] sorted minimizers\n",
      "[M::main::0.132*1.35] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.142*1.33] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.149*1.31] distinct minimizers: 799170 (95.15% are singletons); average occurrences: 1.182; average spacing: 5.292\n",
      "[M::worker_pipeline::8.752*1.01] mapped 2 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Chr22/r1.contigs.fasta\n",
      "[M::main] Real time: 8.757 sec; CPU: 8.816 sec; Peak RSS: 0.169 GB\n"
     ]
    }
   ],
   "source": [
    "# obtain all contigs to one file\n",
    "project_name = \"chr22canu\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa\"\n",
    "\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval \n",
    "\n",
    "!quast.py -a all -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Chr22/r1.contigs.fasta\"\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Chr22/r1.contigs.fasta\" > ./\"$project_name\"/mappings/CANU_alignment.sam\n",
    "!samtools sort ./\"$project_name\"/mappings/CANU_alignment.sam > ./\"$project_name\"/mappings/CANU_alignment.sorted.bam\n",
    "!samtools index -b ./\"$project_name\"/mappings/CANU_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all -o ./r4/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/G4.fa -t 16 ./r4/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-42-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2018-12-20 13:59:31\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/G4.fa ==> G4\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./r4/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2018-12-20 13:59:31\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    G4.fa, length = 5000000, num fragments = 1, GC % = 50.02\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 1749925, L50 = 2, Total length = 4374913, GC % = 50.02, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 13:59:33\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 13:59:34\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 1999988, NA50 = 1749925, NGA50 = 1749925, LA50 = 2, LGA50 = 2\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2018-12-20 13:59:34\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2018-12-20 13:59:35\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2018-12-20 13:59:36\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/r4/quast_eval/quast.log\n",
      "\n",
      "Finished: 2018-12-20 13:59:36\n",
      "Elapsed time: 0:00:05.359214\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.151*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.177*1.30] sorted minimizers\n",
      "[M::main::0.177*1.30] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.190*1.28] mid_occ = 4\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.199*1.26] distinct minimizers: 813698 (85.49% are singletons); average occurrences: 1.146; average spacing: 5.363\n",
      "[M::worker_pipeline::0.670*2.07] mapped 3 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/G4.fa ./r4/contigs/contigs.fasta\n",
      "[M::main] Real time: 0.676 sec; CPU: 1.396 sec; Peak RSS: 0.132 GB\n"
     ]
    }
   ],
   "source": [
    "# obtain all contigs to one file\n",
    "project_name = \"r4\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/G4.fa\"\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "\n",
    "!cat \\\n",
    "'/media/anuradhawick/data/Experiments/Assembly_Graph/Improving_Assembly_Graph/simG4/contigs/contig_1' \\\n",
    "'/media/anuradhawick/data/Experiments/Assembly_Graph/Improving_Assembly_Graph/simG4/contigs/contig_2' \\\n",
    "'/media/anuradhawick/data/Experiments/Assembly_Graph/Improving_Assembly_Graph/simG4/contigs/contig_3' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "\n",
    "!quast.py -a all -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./chr22_contig5/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa -t 16 ../Improving_Assembly_Graph/chr22/contigs/contig_5\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-02 09:47:03\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa ==> chr22_5mb\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ../Improving_Assembly_Graph/chr22/contigs/contig_5 ==> contig_5\n",
      "\n",
      "2019-01-02 09:47:04\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    chr22_5mb.fa, length = 5000000, num fragments = 1, GC % = 45.41\n",
      "  Contig files: \n",
      "    contig_5\n",
      "  Calculating N50 and L50...\n",
      "    contig_5, N50 = 6831, L50 = 1, Total length = 6831, GC % = 36.52, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contig_5 GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/basic_stats/contig_5_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:47:04\n",
      "Running Contig analyzer...\n",
      "  contig_5\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/contigs_report_contig_5.stdout and contigs_report_contig_5.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:47:05\n",
      "Running NA-NGA calculation...\n",
      "  contig_5, Largest alignment = 6831, NA50 = 6831, LA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:47:05\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contig_5\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-02 09:47:06\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-02 09:47:07\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig5/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-02 09:47:07\n",
      "Elapsed time: 0:00:03.456713\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "project_name = \"chr22_contig5\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa\"\n",
    " \n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_4' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 \"../Improving_Assembly_Graph/chr22/contigs/contig_5\"\n",
    "\n",
    "# !minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./chr22_contig4/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa -t 16 ./chr22_contig4/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-02 09:49:53\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa ==> chr22_5mb\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./chr22_contig4/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2019-01-02 09:49:53\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    chr22_5mb.fa, length = 5000000, num fragments = 1, GC % = 45.41\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 29485, L50 = 1, Total length = 29485, GC % = 44.16, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:49:54\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:49:54\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 29485, NA50 = 29485, LA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:49:55\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-02 09:49:55\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-02 09:49:56\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-02 09:49:56\n",
      "Elapsed time: 0:00:03.375383\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "project_name = \"chr22_contig4\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/chr22_5mb.fa\"\n",
    " \n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_4' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "# !minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./chr22_contig4to2/quast_eval/ -r ../Improving_Assembly_Graph/chr22/contigs/contig_2 -t 16 ./chr22_contig4to2/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-02 09:53:01\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Improving_Assembly_Graph/chr22/contigs/contig_2 ==> contig_2\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./chr22_contig4to2/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2019-01-02 09:53:01\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    contig_2, length = 266413, num fragments = 1, GC % = 47.67\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 29485, L50 = 1, Total length = 29485, GC % = 44.16, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:53:02\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:53:02\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 29485, NA50 = 29485, LA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-02 09:53:02\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-02 09:53:02\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-02 09:53:03\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_contig4to2/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-02 09:53:03\n",
      "Elapsed time: 0:00:01.756174\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.019*1.08] collected minimizers\n",
      "[M::mm_idx_gen::0.025*1.55] sorted minimizers\n",
      "[M::main::0.025*1.55] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.026*1.54] mid_occ = 39\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.027*1.53] distinct minimizers: 45547 (95.33% are singletons); average occurrences: 1.105; average spacing: 5.292\n",
      "[M::worker_pipeline::0.034*1.42] mapped 1 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a ../Improving_Assembly_Graph/chr22/contigs/contig_2 ./chr22_contig4to2/contigs/contigs.fasta\n",
      "[M::main] Real time: 0.035 sec; CPU: 0.049 sec; Peak RSS: 0.008 GB\n"
     ]
    }
   ],
   "source": [
    "project_name = \"chr22_contig4to2\"\n",
    "reference_genome = '../Improving_Assembly_Graph/chr22/contigs/contig_2'\n",
    " \n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/chr22/contigs/contig_4' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./chr22_flye_to_canu/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/canutig22.fasta -t 16 /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/flye2_4_5.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-03 16:10:53\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/canutig22.fasta ==> canutig22\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/flye2_4_5.fasta ==> flye2_4_5\n",
      "\n",
      "2019-01-03 16:10:53\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    canutig22.fasta, length = 25466, num fragments = 1, GC % = 50.32\n",
      "  Contig files: \n",
      "    flye2_4_5\n",
      "  Calculating N50 and L50...\n",
      "    flye2_4_5, N50 = 266413, L50 = 1, Total length = 302729, GC % = 47.08, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing flye2_4_5 GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/basic_stats/flye2_4_5_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-03 16:10:54\n",
      "Running Contig analyzer...\n",
      "  flye2_4_5\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/contigs_report_flye2_4_5.stdout and contigs_report_flye2_4_5.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-03 16:10:54\n",
      "Running NA-NGA calculation...\n",
      "  flye2_4_5, Largest alignment = 140, NA50 = None, LA50 = None\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-03 16:10:54\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  flye2_4_5\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-03 16:10:54\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-03 16:10:55\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/chr22_flye_to_canu/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-03 16:10:55\n",
      "Elapsed time: 0:00:02.542194\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.003*1.47] collected minimizers\n",
      "[M::mm_idx_gen::0.005*1.90] sorted minimizers\n",
      "[M::main::0.005*1.89] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.006*1.81] mid_occ = 24\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.006*1.76] distinct minimizers: 4201 (94.14% are singletons); average occurrences: 1.169; average spacing: 5.186\n",
      "[M::worker_pipeline::0.512*1.16] mapped 3 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/canutig22.fasta /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/flye2_4_5.fasta\n",
      "[M::main] Real time: 0.513 sec; CPU: 0.595 sec; Peak RSS: 0.167 GB\n"
     ]
    }
   ],
   "source": [
    "project_name = \"chr22_flye_to_canu\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/canutig22.fasta\"\n",
    " \n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 \"/media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/flye2_4_5.fasta\"\n",
    "!minimap2 -a \"$reference_genome\" \"/media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/canu_contigs_chr22/flye2_4_5.fasta\" > ./\"$project_name\"/mappings/alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Yeast Genome Contig Realignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./yeastgenome/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa -t 16 ./yeastgenome/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-04 13:35:20\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ==> YeastGenome\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./yeastgenome/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2019-01-04 13:35:22\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    YeastGenome.fa, length = 12157105, num fragments = 1, GC % = 38.15\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 2619345, L50 = 2, Total length = 12074401, GC % = 38.17, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:24\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:29\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 5473509, NA50 = 1895493, NGA50 = 1895493, LA50 = 2, LGA50 = 2\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:30\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-04 13:35:34\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-04 13:35:35\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-04 13:35:35\n",
      "Elapsed time: 0:00:14.328828\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.355*1.00] collected minimizers\n",
      "[M::mm_idx_gen::0.578*1.08] sorted minimizers\n",
      "[M::main::0.578*1.08] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.617*1.07] mid_occ = 33\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.644*1.07] distinct minimizers: 2084960 (95.68% are singletons); average occurrences: 1.093; average spacing: 5.336\n",
      "[M::worker_pipeline::3.342*1.83] mapped 18 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ./yeastgenome/contigs/contigs.fasta\n",
      "[M::main] Real time: 3.350 sec; CPU: 6.137 sec; Peak RSS: 0.307 GB\n"
     ]
    }
   ],
   "source": [
    "project_name = \"yeastgenome\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa\"\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_1' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_2' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_3' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_4' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_5' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_6' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_7' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_8' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_9' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_10' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_11' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_12' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_13' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_14' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_15' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_16' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_17' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_18' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all --ambiguity-score 0.8 -o ./yeastgenome_refinedgraph/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa -t 16 ./yeastgenome_refinedgraph/contigs/contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-04 13:35:43\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ==> YeastGenome\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  ./yeastgenome_refinedgraph/contigs/contigs.fasta ==> contigs\n",
      "\n",
      "2019-01-04 13:35:44\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    YeastGenome.fa, length = 12157105, num fragments = 1, GC % = 38.15\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 2619345, L50 = 2, Total length = 12019949, GC % = 38.16, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/basic_stats/contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:46\n",
      "Running Contig analyzer...\n",
      "  contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/contigs_report_contigs.stdout and contigs_report_contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:52\n",
      "Running NA-NGA calculation...\n",
      "  contigs, Largest alignment = 5473509, NA50 = 1895493, NGA50 = 1895493, LA50 = 2, LGA50 = 2\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 13:35:52\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-04 13:35:56\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-04 13:35:57\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastgenome_refinedgraph/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-04 13:35:57\n",
      "Elapsed time: 0:00:14.357697\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "[M::mm_idx_gen::0.360*0.99] collected minimizers\n",
      "[M::mm_idx_gen::0.549*1.13] sorted minimizers\n",
      "[M::main::0.549*1.13] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.582*1.12] mid_occ = 33\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.610*1.12] distinct minimizers: 2084960 (95.68% are singletons); average occurrences: 1.093; average spacing: 5.336\n",
      "[M::worker_pipeline::3.371*1.79] mapped 7 sequences\n",
      "[M::main] Version: 2.14-r886-dirty\n",
      "[M::main] CMD: minimap2 -a /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ./yeastgenome_refinedgraph/contigs/contigs.fasta\n",
      "[M::main] Real time: 3.380 sec; CPU: 6.051 sec; Peak RSS: 0.308 GB\n"
     ]
    }
   ],
   "source": [
    "project_name = \"yeastgenome_refinedgraph\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa\"\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/contigs ] && mkdir \"$project_name\"/contigs\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval\n",
    "\n",
    "!cat \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_1' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_2' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_3' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_4' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_5' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_6' \\\n",
    "'../Improving_Assembly_Graph/yeast/contigs/contig_15' \\\n",
    "> ./\"$project_name\"/contigs/contigs.fasta \n",
    "\n",
    "!quast.py -a all --ambiguity-score 0.8 -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 ./\"$project_name\"/contigs/contigs.fasta\n",
    "\n",
    "!minimap2 -a \"$reference_genome\" ./\"$project_name\"/contigs/contigs.fasta > ./\"$project_name\"/mappings/FLYE_alignment.sam\n",
    "# !samtools sort ./\"$project_name\"/mappings/FLYE_alignment.sam > ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam\n",
    "# !samtools index -b ./\"$project_name\"/mappings/FLYE_alignment.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/anuradhawick/anaconda3/lib/python3.6/site-packages/quast-5.0.2-py3.6.egg/EGG-INFO/scripts/quast.py -a all -o ./yeastcanu/quast_eval/ -r /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa -t 16 /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Yeast/r1.contigs.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.15.0-43-generic-x86_64-with-debian-buster-sid (linux_64)\n",
      "  Python version: 3.6.5\n",
      "  CPUs number: 8\n",
      "\n",
      "Started: 2019-01-04 16:44:53\n",
      "\n",
      "Logging to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/quast.log\n",
      "\n",
      "CWD: /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations\n",
      "Main parameters: \n",
      "  MODE: default, threads: 16, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: all, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa ==> YeastGenome\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  /media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Yeast/r1.contigs.fasta ==> r1.contigs\n",
      "\n",
      "2019-01-04 16:44:54\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    YeastGenome.fa, length = 12157105, num fragments = 1, GC % = 38.15\n",
      "  Contig files: \n",
      "    r1.contigs\n",
      "  Calculating N50 and L50...\n",
      "    r1.contigs, N50 = 8690606, L50 = 1, Total length = 12224431, GC % = 38.16, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/basic_stats/GC_content_plot.pdf\n",
      "  Drawing r1.contigs GC content plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/basic_stats/r1.contigs_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 16:44:55\n",
      "Running Contig analyzer...\n",
      "  r1.contigs\n",
      "  Logging to files /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/contigs_report_r1-contigs.stdout and contigs_report_r1-contigs.stderr...\n",
      "  Aligning contigs to the reference\n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 16:45:00\n",
      "Running NA-NGA calculation...\n",
      "  r1.contigs, Largest alignment = 8683611, NA50 = 8683611, NGA50 = 8683611, LA50 = 1, LGA50 = 1\n",
      "  Drawing cumulative plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-01-04 16:45:01\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  r1.contigs\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-01-04 16:45:03\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-01-04 16:45:04\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/report.html\n",
      "  PDF version (tables and plots) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/report.pdf\n",
      "  Icarus (contig browser) is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/icarus.html\n",
      "  Log is saved to /media/anuradhawick/data/Experiments/Assembly_Graph/Estimating_actual_contig_locations/yeastcanu/quast_eval/quast.log\n",
      "\n",
      "Finished: 2019-01-04 16:45:04\n",
      "Elapsed time: 0:00:11.178189\n",
      "NOTICEs: 3; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "# obtain all contigs to one file\n",
    "project_name = \"yeastcanu\"\n",
    "reference_genome = \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Sample_Genomes/yeast/YeastGenome.fa\"\n",
    "\n",
    "\n",
    "![ -e \"$project_name\" ] && rm -r \"$project_name\"\n",
    "!mkdir \"$project_name\"\n",
    "![ ! -e \"$project_name\"/mappings ] && mkdir \"$project_name\"/mappings\n",
    "![ ! -e \"$project_name\"/quast_eval ] && mkdir \"$project_name\"/quast_eval \n",
    "\n",
    "!quast.py -a all -o ./\"$project_name\"/quast_eval/ -r \"$reference_genome\" -t 16 \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Yeast/r1.contigs.fasta\"\n",
    "\n",
    "# !minimap2 -a \"$reference_genome\" \"/media/anuradhawick/data/Experiments/Assembly_Graph/Assembly/Canu/Out Yeast/r1.contigs.fasta\" > ./\"$project_name\"/mappings/CANU_alignment.sam"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
